Implement FieldSqrt hint

mmk-1 opened the pull request:
Resolves issue #178

cicr99 reviewed (commented):
Other than the comment about the test, looks good to me

mmk-1 commented:
@cicr99 I added a new test cases for when the input is quadratic non-residue.
I also added a comments for the `Legendre` method with a link that gives the mathematical reasoning how it works. What do you think?
